### PR TITLE
Live tests: GHA to run both validation & regression tests

### PR DESCRIPTION
**Pull request by clnoll**
## What
Add the ability to run all live tests in CI.

## How
Updates `airbyte-ci` to have a new command for running all validation + regression tests. Ultimately we may want to remove the regression tests altogether.

Also adds a GHA for running validation + regression tests.

Closes https://github.com/airbytehq/airbyte-internal-issues/issues/7854.

### CODE REVIEW

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py:593]: 📚 The explanation is a bit far from the weird stuff were doing. We should move this to another small function and likely discuss what were doing, and why, for future maintainers

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py:295]: ❓ Do we have any reason to not run these tests we need to plan for?

**Review by bnchrch — COMMENTED** [review]
Just a few questions!

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py
Comment (clnoll) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py:295]: By default, [they'll always be skipped unless specifically selected](https://github.com/airbytehq/airbyte/pull/38816/files#diff-3a2feb8da6d3b4507f7a42303e54de4ac916308de4501f2e77d6ee46c97b45edR118). Whenever we're ready for broader automation we'll want to talk about the frequency.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py
Comment (clnoll) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py]: Good call-out. I've moved it into its own method and enhanced the docstring.

File: .github/workflows/validation_tests.yml
Comment (natikgadzhi) [inline .github/workflows/validation_tests.yml:20]: Should connection_id be optional if we're supporting live validation tests too?

**Review by natikgadzhi — COMMENTED** [review]
I'm not 100% certain, but I would love us to pick this up and merge this without waiting for 2 weeks for Catherine to get back.

I'd love to get to deterministic output of these tests (pass/fail) for validation tests.

File: .github/workflows/validation_tests.yml
Comment (alafanechere) [reply .github/workflows/validation_tests.yml:20]: We could set `auto` as a default. I think @clnoll original intent was to make users very intentional about which connection is used for testing.

File: .github/workflows/validation_tests.yml
Comment (alafanechere) [reply .github/workflows/validation_tests.yml:45]: In it's current state this worklow can't run on fork as:
* It is triggered by worfklow dispatch event, the checked out repo is the airbyte one
* There's no input to make this worfklow run on a different repo

I think it's fair to port it to the community_ci.yml workflow once we validate it can automatically run these test suite on our own PRs. Which is not the scope of this current change.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py:333]: as `should_read_with_state`  or `test_evaluation_mode` attributes are not set in `__init__` I'd suggest making this function static and make the caller pass these parameters.
We have no guarantee that ` self.test_evaluation_mode` is set...

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py:308]: As `LiveTests` is a "macro" test suite composed of `ValidationTests` and `RegressionTests` I would suggest making it "abstract" and implement a specific `ValidationTests` step...

In it's current form I can't see how this step is execute a different thing than `RegressionTests` as the test input directory remains `src/live_tests/regression_tests`

**Review by alafanechere — CHANGES_REQUESTED** [review]
I don't think I get how `LiveTests` is different from `RegressionTests`.
I'd have expected two separate concrete steps (`RegressionTests` and `ValidationTests`) which would inherit from `LiveTests` if we want to share some execution logic between the two.
Having two different Step classes would help in differentiating the two test suites, their purpose, and in conditionally controlling their execution.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py
Comment (clnoll) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py]: > As LiveTests is a "macro" test suite composed of ValidationTests and RegressionTests I would suggest making it "abstract" and implement a specific ValidationTests step...

Done.

> In it's current form I can't see how this step is execute a different thing than RegressionTests as the test input directory remains src/live_tests/regression_tests

Oops, that was an oversight on my part. Updated the input directory to `src/live_tests`.

**Review by clnoll — COMMENTED** [review]
@alafanechere thanks for the comments. I've added a new ValidationTest step, in addition to the LiveTest step that runs validation + regression. 

For now I'm holding off on a GHA specifically for validation tests, but will add one if it feels like it will be needed. It feels likely that we'll want to deprecate the regression tests GHA instead, and opt to always run regression + validation together.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py:306]: We have a test step granularity decision to make.
My opinion is that we should have two distincts steps: `ValidationTests` and `RegressionTests`.
They are both sub-suites of live tests s but we might orchestrate these test suites differently according to the connector state (we might allow CI to pass if validation fails for the same reason as on master, but fail it if regression test fail etc.).

In others words: lets not make airbyte-ci aware that the live test suite can be used to run both regression and validation.

```suggestion
            StepToRun(
                id=CONNECTOR_TEST_STEP_ID.CONNECTOR_VALIDATION_TESTS,
                step=ValidationTests(context),
                args=lambda results: {"connector_under_test_container": results[CONNECTOR_TEST_STEP_ID.BUILD].output[LOCAL_BUILD_PLATFORM]},
                depends_on=[CONNECTOR_TEST_STEP_ID.BUILD],
            ),
       
```

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py:476]:
```suggestion
```
This is already an abstract property of `Step`

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py:452]: nit: I'm not sure if this path refers to a local one or an in-container one. A comment or a rename might help.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py:118]: nit: might be helpful to keep this list in a constant named something like `STEPS_SKIPPED_BY_DEFAULT`

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py:703]: nit: I'd recommend returning the a list and joining it at call time. Just because the usual `with_exec` input is a list. (the fact we join it is an implementation detail at the container execution logic level)

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py:722]: these `or ""` look strange to me, shall we remove the option name for the list if attributes are falsy?

File: airbyte-ci/connectors/live-tests/src/live_tests/commons/json_schema_helper.py
Comment (alafanechere) [inline airbyte-ci/connectors/live-tests/src/live_tests/commons/json_schema_helper.py:1]: Can you share what motivated these changes? They do not look related to the scope of the PR.

File: .github/workflows/validation_tests.yml
Comment (alafanechere) [inline .github/workflows/validation_tests.yml:89]:
```suggestion
      - name: Run  Validation Tests [WORKFLOW DISPATCH]
```

File: .github/workflows/validation_tests.yml
Comment (alafanechere) [inline .github/workflows/validation_tests.yml:105]:
```suggestion
          subcommand: connectors ${{ env.USE_LOCAL_CDK_FLAG }} --name ${{ github.event.inputs.connector_name }} test --only-step connector_validation_tests --connector_validation_tests.connection-id=${{ github.event.inputs.connection_id }} --connector_validation_tests.pr-url=${{ github.event.inputs.pr_url }} ${{ env.STREAM_PARAMS }} --connector_validation_tests.test-evaluation-mode=diagnostic
```

**Review by alafanechere — CHANGES_REQUESTED** [review]
Just requesting change because I believe we always want to run Validation and Regression separately.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py:306]: > but the simplest way to reuse the test artifacts 

We could create a  dagger cache volume on the artifact path which would make sure artifacts are shared between steps.

> avoid pytest rerunning the commands

Do you mean the connector commands?
The connector runner uses dagger under the hood, so if the command inputs did not change between the step their output would be cache. (the behavior has to be tested).

But I hear your :) Then  lets make a single `LiveTest` step and control what test suite run in it (validation/regression) via step arguments.

This delays the complexity of decoupling the test suites in separate steps if we ever want to make our CI logic more granular. But I'm fine with it.

**Review by alafanechere — APPROVED** [review]

**Pull request merged**
